cabal-install/Distribution/Client/ProjectPlanning/Types.hs

```diff
@@ -91,16 +107,155 @@ data ElaboratedSharedConfig
+elabReplTarget (ElabComponent comp)
+    | Just cname <- elabComponentName comp
+    = fmap (ComponentTarget cname) $ elabComponentReplTarget comp
+    | otherwise = Nothing
+
+elabBuildHaddocks :: ElaboratedConfiguredPackage -> Bool
+elabBuildHaddocks (ElabPackage pkg) = pkgBuildHaddocks pkg
+elabBuildHaddocks (ElabComponent comp) = elabComponentBuildHaddocks comp
+
+getElaboratedPackage :: ElaboratedConfiguredPackage -> ElaboratedPackage
+getElaboratedPackage (ElabPackage pkg) = pkg
+getElaboratedPackage (ElabComponent comp) = elabComponentPackage comp
+
+instance Binary ElaboratedConfiguredPackage
+
+instance Package ElaboratedConfiguredPackage where
+  packageId (ElabPackage pkg) = packageId pkg
+  packageId (ElabComponent comp) = packageId comp
+
+instance HasUnitId ElaboratedConfiguredPackage where
+  installedUnitId (ElabPackage pkg) = installedUnitId pkg
+  installedUnitId (ElabComponent comp) = installedUnitId comp
+
+instance HasConfiguredId ElaboratedConfiguredPackage where
+  configuredId (ElabPackage pkg)    = configuredId pkg
+  configuredId (ElabComponent comp) = configuredId comp
+
+-- | Some extra metadata associated with an
+-- 'ElaboratedConfiguredPackage' which indicates that the "package"
+-- in question is actually a single component to be built.  Arguably
+-- it would be clearer if there were an ADT which branched into
+-- package work items and component work items, but I've structured
+-- it this way to minimize change to the existing code (which I
+-- don't feel qualified to rewrite.)
+data ElaboratedComponent
+   = ElaboratedComponent {
+    -- | The munged package identifier of this component.
+    -- Invariant that the 'Version' of this 'PackageId'
+    -- matches that in 'ElaboratedPackage'.
+    elabComponentPackageId :: PackageId,
+    -- | The name of the component to be built
+    elabComponent :: CD.Component,
+    -- | The name of the component to be built.  Nothing if
+    -- it's a setup dep.
+    elabComponentName :: Maybe ComponentName,
+    -- | The ID of the component to be built
+    elabComponentId :: UnitId,
+    -- | Dependencies of this component
+    elabComponentDependencies :: [ConfiguredId],
+    -- | The order-only dependencies of this component; e.g.,
+    -- if you depend on an executable it goes here.
+    elabComponentExeDependencies :: [UnitId],
+    -- | The 'ElaboratedPackage' this component came from
+    elabComponentPackage :: ElaboratedPackage,
+    -- | What in this component should we build (TRANSIENT, see 'pkgBuildTargets')
+    elabComponentBuildTargets :: [SubComponentTarget],
+    -- | Should we REPL this component (TRANSIENT, see 'pkgReplTarget')
+    elabComponentReplTarget :: Maybe SubComponentTarget,
+    -- | Should we Haddock this component (TRANSIENT, see 'pkgBuildHaddocks')
+    elabComponentBuildHaddocks :: Bool
+    -- NB: Careful, if you add elabComponentInstallDirs, need
+    -- to adjust 'packageHashInputs'!!!
+   }
```

Comment:
There is actually very little this data type has in common with `ElaboratedPackage`. For example, `elabComponentDependencies` has the type `[ConfiguredId]`, whereas `pkgDependencies :: ComponentDeps [ConfiguredId]` (since it needs separate deps per component.) So the most you could share is the `elabComponentId` and `elabComponentPackageId` which seems hardly worth it. Similarly, `elabComponentBuildTargets :: [SubComponentTarget]`, but has `pkgBuildTargets :: [ComponentTarget]`, indicating the fact that targets for a package range over multiple components, whereas for a component, the component is already known and it can only be a subtarget.

Originally, I prototyped this patch by giving `ElaboratedPackage` a subfield containing extra component-y stuff. Eventually I decided this was a bad idea.